File: sortAlgorithim/Java/MergeSort.java
package sortAlgorithm;

	import java.util.Arrays;
	
	public class MergeSort {
		public  static int[] merge(int[]left,int[]right){
			int index=0,l=0,r=0,min=0;
			int length=left.length+right.length;
			int[] tmp = new int[length];
			for(;index<length;){
				if(l<left.length&&r<right.length){
					if(left[l]<=right[r]){
						min = left[l];
						l++;
					}else{
						min = right[r];
						r++;
					}
					tmp[index] = min;
					index++;
				}
				
				if(l==left.length&&r<right.length){
					for(;r<right.length;r++){
						tmp[index] = right[r];
						index++;
					}
				}
				
				if(r==right.length&&l<left.length){
					for(;l<left.length;l++){
					tmp[index] = left[l];
					index++;
					}
				}
			}
			return tmp;
		}
		public static int[] mergeSort(int[] arr){
			if(arr.length<=1)return arr;
			int mid = arr.length/2;
			int[] left = Arrays.copyOfRange(arr, 0,mid);
			int[] right = Arrays.copyOfRange(arr, mid,arr.length);
			return merge(mergeSort(left),mergeSort(right));
		}
		public static void show(int[] nums) {
			for(int i =0;i<nums.length;i++) {
				System.out.printf("%d ", nums[i]);
			}
			System.out.println();
		}
		public static void main(String[] args) {
			// TODO Auto-generated method stub
			int[] nums = {5,6,3,1,8,7,2,4};
			System.out.println("排序前：");
			show(nums);
			int[] numsSorted = mergeSort(nums);
			System.out.println("排序后：");
			show(numsSorted);
		}
	}

File: sortAlgorithim/Go/MergeSort.go
package main

import "fmt"

func show(nums []int){
	for _,num:=range nums{
		fmt.Printf("%d ",num)
	}
	fmt.Println()
}
func merge(left []int,right []int)[]int{
	index,l,r,min,length:=0,0,0,0,len(left)+len(right)
	tmp := make([]int,length)
	for ;index<length;{
		if l<len(left)&&r<len(right) {
			if left[l]<=right[r]{
				min = left[l]
				l++
			}else{
				min = right[r]
				r++
			}
			tmp[index] = min
			index++
		}
		if l==len(left)&&r<len(right) {
			for;r<len(right);r++{
				tmp[index] = right[r]
				index++
			}
		}

		if r==len(right)&&l<len(left){
			for;l<len(left);l++{
			tmp[index] = left[l]
			index++
			}
		}
	}
	return tmp
}
func mergeSort(nums []int)[]int{
	if len(nums)<=1{
		return nums
	}
	mid:=(int)(len(nums)/2)
	left := nums[0:mid]
	right:= nums[mid:len(nums)]

	return merge(mergeSort(left),mergeSort(right))
}
func main(){
	nums :=[]int{5,6,3,1,8,7,2,4}
	fmt.Println("排序前：")
	show(nums)
	numsSorted := mergeSort(nums)
	fmt.Println("排序后：")
	show(numsSorted)
}
